plugins/ingest/src/main/java/org/elasticsearch/plugin/ingest/PipelineStoreBootstrapper.java

```diff
@@ -0,0 +1,101 @@
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.plugin.ingest;
+
+import org.elasticsearch.cluster.ClusterChangedEvent;
+import org.elasticsearch.cluster.ClusterService;
+import org.elasticsearch.cluster.ClusterStateListener;
+import org.elasticsearch.cluster.routing.IndexRoutingTable;
+import org.elasticsearch.common.component.AbstractLifecycleComponent;
+import org.elasticsearch.common.inject.Inject;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.common.unit.TimeValue;
+import org.elasticsearch.gateway.GatewayService;
+import org.elasticsearch.threadpool.ThreadPool;
+
+import java.io.IOException;
+
+public final class PipelineStoreBootstrapper extends AbstractLifecycleComponent implements ClusterStateListener {
+
+    private final ThreadPool threadPool;
+    private final PipelineStore pipelineStore;
+    private final ClusterService clusterService;
+    private final TimeValue pipelineUpdateInterval;
+
+    @Inject
+    public PipelineStoreBootstrapper(Settings settings, ThreadPool threadPool, PipelineStore pipelineStore, ClusterService clusterService) {
+        super(settings);
+        this.threadPool = threadPool;
+        this.pipelineStore = pipelineStore;
+        this.clusterService = clusterService;
+        this.pipelineUpdateInterval = settings.getAsTime("ingest.pipeline.store.update.interval", TimeValue.timeValueMinutes(30));
+        clusterService.add(this);
+    }
+
+    @Override
+    public void clusterChanged(ClusterChangedEvent event) {
+        if (event.state().blocks().hasGlobalBlock(GatewayService.STATE_NOT_RECOVERED_BLOCK) == false) {
+            boolean canInitialize;
+            if (event.state().getMetaData().hasConcreteIndex(PipelineStore.INDEX) == false) {
+                // it will be ready when auto create index kicks in before the first pipeline doc gets added
+                canInitialize = true;
+            } else {
+                IndexRoutingTable routingTable = event.state().getRoutingTable().index(PipelineStore.INDEX);
+                canInitialize = routingTable.allPrimaryShardsActive();
+            }
+
+            if (canInitialize) {
+                pipelineStore.initialize();
+                scheduleBackgroundUpdater();
+                clusterService.remove(this);
+            }
+        }
+    }
+
+    @Override
+    protected void doStart() {
+    }
+
+    @Override
+    protected void doStop() {
+    }
+
+    @Override
+    protected void doClose() {
+        try {
+            pipelineStore.close();
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    void scheduleBackgroundUpdater() {
```

Comment:
> all I want here is a mechanism that always works.

we have this common goal, I agree that not having a background task would be cleaner, but I wonder how we can address all the different problems of a distributed system (a node loses connectivity and comes back up, the search for pipelines returns an error e.g. shard not available etc.) with a sync approach. Sounds like moving completely to sync at the moment means that we will potentially not find pipelines or execute not up-to-date pipelines... Maybe we should identify changes that have to be made to core so that everything can be made sync, but in the meantime, this PR is already a big improvement as it makes things testable and at least partially sync.